lib/concurrent/timer_task2.rb

```diff
@@ -0,0 +1,297 @@
+  #     raise Concurrent::TimeoutError.new if cancellation.cancelled?
+  #   end
+  #   task.execute
+  #
+  #   3.times do
+  #     observe.call(channel)
+  #   end
+  #   #=> (false) Execution timed out
+  #   #=> (false) Execution timed out
+  #   #=> (false) Execution timed out
+  #   task.shutdown
+  #   channel.pop
+  #
+  #   task = Concurrent::TimerTask2.new(execution_interval: 1, channel: channel){ raise StandardError }
+  #   task.execute
+  #
+  #   3.times do
+  #     observe.call(channel)
+  #   end
+  #   #=> (false) Execution failed with error StandardError
+  #   #=> (false) Execution failed with error StandardError
+  #   #=> (false) Execution failed with error StandardError
+  #   task.shutdown
+  #   channel.pop
+  #
+  # @see http://docs.oracle.com/javase/7/docs/api/java/util/TimerTask.html
+  class TimerTask2
+
+    # Default `:execution_interval` in seconds.
+    EXECUTION_INTERVAL = 60
+    # Default `:timeout_interval` in seconds.
+    TIMEOUT_INTERVAL = 30
+
+    # Create a new TimerTask2 with the given task and configuration.
+    #
+    # @!macro timer_task_initialize
+    #   @param [Hash] opts the options defining task execution.
+    #   @option opts [Integer] :execution_interval number of seconds between
+    #     task executions (default: EXECUTION_INTERVAL)
+    #   @option opts [Integer] :timeout_interval number of seconds a task can
+    #     run before it is considered to have failed (default: TIMEOUT_INTERVAL)
+    #   @option opts [Boolean] :run_now Whether to run the task immediately
+    #     upon instantiation or to wait until the first execution_interval
+    #     has passed (default: false)
+    #   @option opts [Symbol] :reschedule when to schedule next execution relative,
+    #     relative to a current one, `:before` or `:after` (default: :after)
+    #   @option opts [Promises::Channel, nil] :channel if given, execution results
+    #     will be pushed into the channel (default: nil)
+    #
+    #   @raise ArgumentError when no block is given.
+    #
+    #   @yield to the block after :execution_interval seconds have passed since
+    #     the last yield
+    #   @yieldparam task a reference to the `TimerTask` instance so that the
+    #     block can control its own lifecycle. Necessary since `self` will
+    #     refer to the execution context of the block rather than the running
+    #     `TimerTask`.
+    #   @yieldparam cancellation a cancellation object representing the joined cancellation
+    #     of the timer task and this run's timeout
+    #
+    #   @return [TimerTask] the new `TimerTask`
+    def initialize(opts = {}, &task)
+      raise ArgumentError.new('no block given') unless block_given?
+      raise ArgumentError.new('reschedule must be either :before or :after') unless [nil, :before, :after].include?(opts[:reschedule])
+      @execution_interval = AtomicReference.new nil
+      @timeout_interval = AtomicReference.new nil
+      @cancellation = AtomicReference.new nil
```

Comment:
Please use Capitalized names for the instance variables, since they are final.

Reply:
Should they use capitalized snake case or camel/pascal case?

Reply:
`@TimeoutInterval` same as class, thanks